### ISSUE TITLE
Trying To Remove Api Key & Get Error: Output 'GraphQLAPIKeyOutput' not found in stack

### ISSUE THREAD

**Issue by lr001dev** [aws-amplify/amplify-cli#6143]
**Describe the bug**
So when using amplify push after I have tried removing the use of API Key in graphql api using amplify update api I receive this error: Output 'GraphQLAPIKeyOutput' not found in stack

I have a lambda function that was setup to access this api. The above error seems to be related to my Serverless Express Lambda function.  I've tried different things and can't seem to figure out to how to get this to work without having to rebuild the api. 

**Amplify CLI Version**
4.30.0

**To Reproduce**
Amplify Update Api
Remove Api Key
Amplify Push

**Expected behavior**
Should be able to push with no error.

**Screenshots**

**Desktop (please complete the following information):**
 - OS: Mac
 - Node Version: v12.18.3

**Additional context**
Basically I have been using API key until now as we are switching to authenticated requests through cognito user pool and groups. So just trying to make the switch and can't.

**Comment by attilah** [comment]
@lr001dev please follow the issue template and add all the missing information about versions, and repro steps to be able to help you better.

**Comment by kaustavghosh06** [comment]
Closing due to lack of response.

**Comment by lr001dev** [comment]
@kaustavghosh06 I just updated the issue with the templated info, please reopen. Thanks!

**Comment by lr001dev** [comment]
Hello again, wondering if someone connect with me on this... This issue is holding up our development. Thank you!

**Comment by ammarkarachi** [comment]
@lr001dev I will mark this as a bug, You would need to remove the parameter in your function cfn at path `amplify/backend/function/<resource-name>/<function-name>-cloudformation-template.json`

```json
    "api<project-name>GraphQLAPIKeyOutput": {
      "Type": "String",
      "Default": "apitagsGraphQLAPIKeyOutput"
    }
```

**Comment by lr001dev** [comment]
Ok, removed the lines which appeared twice in the template file.. one instance with Default and the other with a Ref property. Then ran amplify update api and selected my new auth settings, coginito and iam. Ran amplify push and still received: `UPDATE_FAILED               functiongroovesapi                        AWS::CloudFormation::Stack Sat Jan 23 2021 18:05:36 GMT-0800 (Pacific S
tandard Time) Output 'GraphQLAPIKeyOutput' not found in stack 'arn:aws:cloudformation:us-west-2:907138769176:stack/amplify-groovestreambacken
d-develop-14321-apigrooveql-13UPZZBG5VSKF/a92d64b0-fa55-11ea-bd2d-0239b11280c5'`

**Comment by lr001dev** [comment]
Any other thoughts on this?

**Comment by jhockett** [comment]
@lr001dev can you share the function's cloudformation template?

**Comment by lr001dev** [comment]
I can, how should I send @jhockett ?

**Comment by jhockett** [comment]
@lr001dev copy + paste it here or you can email it to amplify-cli@amazon.com. If you choose email, please include the issue number in the subject. Please redact anything sensitive before sharing either way.

**Comment by lr001dev** [comment]
I removed lambda role names as indicated by `""`

`{
  "AWSTemplateFormatVersion": "2010-09-09",
  "Description": "Lambda Function resource stack creation using Amplify CLI",
  "Parameters": {
    "CloudWatchRule": {
      "Type": "String",
      "Default": "NONE",
      "Description": " Schedule Expression"
    },
    "env": {
      "Type": "String"
    },
    "apigrooveqlGraphQLAPIIdOutput": {
      "Type": "String",
      "Default": "apigrooveqlGraphQLAPIIdOutput"
    },
    "apigrooveqlGraphQLAPIEndpointOutput": {
      "Type": "String",
      "Default": "apigrooveqlGraphQLAPIEndpointOutput"
    }
  },
  "Conditions": {
    "ShouldNotCreateEnvResources": {
      "Fn::Equals": [
        {
          "Ref": "env"
        },
        "NONE"
      ]
    }
  },
  "Resources": {
    "LambdaFunction": {
      "Type": "AWS::Lambda::Function",
      "Metadata": {
        "aws:asset:path": "./src",
        "aws:asset:property": "Code"
      },
      "Properties": {
        "Handler": "index.handler",
        "FunctionName": {
          "Fn::If": [
            "ShouldNotCreateEnvResources",
            "groovesapi",
            {
              "Fn::Join": [
                "",
                [
                  "groovesapi",
                  "-",
                  {
                    "Ref": "env"
                  }
                ]
              ]
            }
          ]
        },
        "Environment": {
          "Variables": {
            "ENV": {
              "Ref": "env"
            },
            "REGION": {
              "Ref": "AWS::Region"
            },
            "API_GROOVEQL_GRAPHQLAPIIDOUTPUT": {
              "Ref": "apigrooveqlGraphQLAPIIdOutput"
            },
            "API_GROOVEQL_GRAPHQLAPIENDPOINTOUTPUT": {
              "Ref": "apigrooveqlGraphQLAPIEndpointOutput"
            }
          }
        },
        "Role": {
          "Fn::GetAtt": [
            "LambdaExecutionRole",
            "Arn"
          ]
        },
        "Runtime": "nodejs12.x",
        "Layers": [],
        "Timeout": "25",
        "Code": {
          "S3Bucket": "amplify-groovestreambackend-develop-14321-deployment",
          "S3Key": "amplify-builds/groovesapi-6953734939335862514b-build.zip"
        }
      }
    },
    "LambdaExecutionRole": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "RoleName": {
          "Fn::If": [
            "ShouldNotCreateEnvResources",
            "",
            {
              "Fn::Join": [
                "",
                [
                  "",
                  "-",
                  {
                    "Ref": "env"
                  }
                ]
              ]
            }
          ]
        },
        "AssumeRolePolicyDocument": {
          "Version": "2012-10-17",
          "Statement": [
            {
              "Effect": "Allow",
              "Principal": {
                "Service": [
                  "lambda.amazonaws.com"
                ]
              },
              "Action": [
                "sts:AssumeRole"
              ]
            }
          ]
        }
      }
    },
    "lambdaexecutionpolicy": {
      "DependsOn": [
        "LambdaExecutionRole"
      ],
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyName": "lambda-execution-policy",
        "Roles": [
          {
            "Ref": "LambdaExecutionRole"
          }
        ],
        "PolicyDocument": {
          "Version": "2012-10-17",
          "Statement": [
            {
              "Effect": "Allow",
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents"
              ],
              "Resource": {
                "Fn::Sub": [
                  "arn:aws:logs:${region}:${account}:log-group:/aws/lambda/${lambda}:log-stream:*",
                  {
                    "region": {
                      "Ref": "AWS::Region"
                    },
                    "account": {
                      "Ref": "AWS::AccountId"
                    },
                    "lambda": {
                      "Ref": "LambdaFunction"
                    }
                  }
                ]
              }
            }
          ]
        }
      }
    },
    "AmplifyResourcesPolicy": {
      "DependsOn": [
        "LambdaExecutionRole"
      ],
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyName": "amplify-lambda-execution-policy",
        "Roles": [
          {
            "Ref": "LambdaExecutionRole"
          }
        ],
        "PolicyDocument": {
          "Version": "2012-10-17",
          "Statement": [
            {
              "Effect": "Allow",
              "Action": [
                "appsync:Create*",
                "appsync:StartSchemaCreation",
                "appsync:GraphQL",
                "appsync:Get*",
                "appsync:List*",
                "appsync:Update*"
              ],
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:aws:appsync:",
                      {
                        "Ref": "AWS::Region"
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId"
                      },
                      ":apis/",
                      {
                        "Ref": "apigrooveqlGraphQLAPIIdOutput"
                      },
                      "/*"
                    ]
                  ]
                }
              ]
            }
          ]
        }
      }
    }
  },
  "Outputs": {
    "Name": {
      "Value": {
        "Ref": "LambdaFunction"
      }
    },
    "Arn": {
      "Value": {
        "Fn::GetAtt": [
          "LambdaFunction",
          "Arn"
        ]
      }
    },
    "Region": {
      "Value": {
        "Ref": "AWS::Region"
      }
    },
    "LambdaExecutionRole": {
      "Value": {
        "Ref": "LambdaExecutionRole"
      }
    }
  }
}
`

**Comment by lr001dev** [comment]
Any love on this?

**Comment by lr001dev** [comment]
@jhockett Possible to get an update on this issue? Any info would be helpful. Thank you!

**Comment by aprilmintacpineda** [comment]
Also having the same issue now. I haven't touched this project for a while, the api key must have expired and now I'm trying to push using CI/CD and it just keeps on failing, I'm tried to set `"CreateAPIKey": 0` on `parameters.json` and this error appeared.

**Comment by SwaySway** [comment]
@lr001dev Do you still want to grant the function access to the API? If so we merged a PR which warns to reselect API in order to update the access to the API.

PR: https://github.com/aws-amplify/amplify-cli/pull/4923

**Comment by aprilmintacpineda** [comment]
I think the error is that the `apiKey` is not yet generated but the function depends on it. I think remove the API from resources of the function.

**Comment by lr001dev** [comment]
> @lr001dev Do you still want to grant the function access to the API? If so we merged a PR which warns to reselect API in order to update the access to the API.
> 
> PR: https://github.com/aws-amplify/amplify-cli/pull/4923

Ideally yes...

**Comment by lr001dev** [comment]
@SwaySway So re-adding the Api Key for the api I was trying to initially removing from.,, then removing it again, then going to the function that is dependent on the api and running amplify update function and re-added the dependency did indeed fix the error. I did see the tip in the console letting me know to re-add the dependency. So that worked!

**Comment by SwaySway** [comment]
Closing this issue - should you experience any issues with re-assigning function permission to api please comment below.
